controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java

```diff
@@ -321,4 +330,97 @@ protected String getRequirementName(OperationContext context, Resource resource,
         }
     }
 
+    /**
+     * {@link CapabilityReferenceRecorder} that determines the dependent and required capability from the
+     * {@link PathAddress} of the resource registration.
+     */
+    class ResourceCapabilityReferenceRecorder implements CapabilityReferenceRecorder {
+
+        private final Function<PathAddress, String[]> dynamicDependentNameMapper;
+        private final Function<PathAddress, String[]> dynamicRequirementNameMapper;
+        private final String baseRequirementName;
+        private final String baseDependentName;
+
+        public ResourceCapabilityReferenceRecorder(Function<PathAddress, String[]> dynamicDependentNameMapper, String baseDependentName, Function<PathAddress, String[]> dynamicRequirementNameMapper, String baseRequirementName) {
+            this.dynamicDependentNameMapper = dynamicDependentNameMapper;
+            this.dynamicRequirementNameMapper = dynamicRequirementNameMapper;
+            this.baseRequirementName = baseRequirementName;
+            this.baseDependentName = baseDependentName;
+        }
+
+        public ResourceCapabilityReferenceRecorder(Function<PathAddress, String[]> dynamicDependentNameMapper, String baseDependentName, String baseRequirementName) {
+            this(dynamicDependentNameMapper, baseDependentName, null, baseRequirementName);
```

Comment:
Why would anyone use this constructor (i.e. a null dynamicRequirementNameMapper)?
If a capability requires a statically named requirement, it would define this requirement via RuntimeCapability.addRequirement(...).  There would be no reason to create a separate capability reference recorder.

Reply:
I agree